src/query/api/v1/handler/prometheus/remote/write.go

```diff
@@ -555,6 +572,84 @@ func (h *PromWriteHandler) forward(
 	return nil
 }
 
+func (h *PromWriteHandler) maybeLogSeriesWithLongestLabels(logger *zap.Logger, timeSeries []prompb.TimeSeries) {
+	if len(timeSeries) == 0 {
+		return
+	}
+
+	// Read the value and check if time series should be logged.
+	h.numLiteralIsTooLongLock.RLock()
+	shouldLog := h.numLiteralIsTooLong < maxLiteralIsTooLongLogCount
+	h.numLiteralIsTooLongLock.RUnlock()
+	if !shouldLog {
+		return
+	}
+
+	// Read and update the value if the time series still should be logged.
+	h.numLiteralIsTooLongLock.Lock()
+	shouldLog = h.numLiteralIsTooLong < maxLiteralIsTooLongLogCount
+	if shouldLog {
+		h.numLiteralIsTooLong++
+	}
+	h.numLiteralIsTooLongLock.Unlock()
+	if !shouldLog {
+		return
+	}
+
+	var (
+		longestLabelLength = 0
+		longestLabelIdx    = 0
+	)
+	for i, ts := range timeSeries {
+		for _, l := range ts.Labels {
+			if len(l.Name) > longestLabelLength {
+				longestLabelLength = len(l.Name)
+				longestLabelIdx = i
+			}
+			if len(l.Value) > longestLabelLength {
+				longestLabelLength = len(l.Value)
+				longestLabelIdx = i
+			}
+		}
+	}
+
+	const (
+		literalLengthThreshold = 30
+		totalLengthThreshold   = 1000
+	)
+
+	labelSummary := labelsToSummaryString(timeSeries[longestLabelIdx].Labels, literalLengthThreshold, totalLengthThreshold)
+	logger.Warn("series with too long literal", zap.String("labelSummary", labelSummary))
```

Comment:
Can we log just a long enough prefix of the id which would reveal any extremely long labels?

Reply:
https://github.com/m3db/m3/pull/3801/commits/6625d7d7a6893fc559382c85dda3e36a5bfc8e17